### ISSUE TITLE
Computed columns seem unstable

### ISSUE THREAD

**Issue by olavloite** [GoogleCloudPlatform/cloud-spanner-emulator#23]
Support for generated columns were added in release 1.2.0 along with a number of other features (thank you for that!), but the computed columns seem to trigger some problems some times.

Using the following table:

```sql
CREATE TABLE Singers (
  SingerId  INT64 NOT NULL,
  FirstName STRING(200),
  LastName  STRING(200) NOT NULL,
  FullName  STRING(400) NOT NULL AS (COALESCE(FirstName || ' ', '') || LastName) STORED,
  BirthDate DATE,
  Picture   BYTES(MAX),
) PRIMARY KEY (SingerId);
```

Then trying to insert a row using the following DML statement:

```sql
INSERT INTO Singers (SingerId, FirstName, LastName) VALUES (@id, @firstName, @lastName)
```
(All inserted values are non-null and within the storage limitations)

Fails on the emulator. Sometimes the emulator crashes, sometimes it generates an error like the one below:

```
    Google.Cloud.Spanner.Data.SpannerException : An unknown error occurred.
    ---- Grpc.Core.RpcException : Status(StatusCode="Unknown", Detail="Unexpected error in RPC handling", DebugException="Grpc.Core.Internal.CoreErrorDetailException: {"created":"@1613582918.298000000","description":"Error received from peer ipv6:[::1]:9010","file":"T:\src\github\grpc\workspace_csharp_ext_windows_x64\src\core\lib\surface\call.cc","file_line":1062,"grpc_message":"Unexpected error in RPC handling","grpc_status":2}")
```

Changing the computed column to a simpler definition, such as for example ` FullName STRING(400) AS (LastName) STORED`, will allow the above DML statement to succeed sometimes, but after 3-4 attempts, it will start failing on my environment.

**Comment by chao-tian** [comment]
Thanks for reporting the issue! 

I tried with your schema but could reproduce the crashes. Do you know if the error is triggered by some specific SQL parameters, or it is triggered randomly? Does it happen without the generated columns? If possible, could you share the parameters that triggered the crashes?

**Comment by olavloite** [comment]
It seems to be triggered randomly, and on my local system that is very often. The specific example data model that I gave above fails every time for me. When I change the computed column to a simpler expression, such as for example `FullName STRING(400) AS (LastName) STORED`, it will take longer to fail (but still fail within something like 5-10 statements). The following example application fails each time I execute it:

```java
  public static void main(String[] args) throws Exception {
    try (Spanner spanner = SpannerOptions.newBuilder().setProjectId("test-project")
        .build().getService()) {
      try {
        spanner.getInstanceAdminClient().createInstance(
            InstanceInfo.newBuilder(InstanceId.of("test-project", "test-instance"))
            .setDisplayName("test")
            .setNodeCount(1)
            .setInstanceConfigId(InstanceConfigId.of("test-project", "emulator-config"))
            .build()
            ).get();
      } catch (ExecutionException ee) {
        SpannerException e = (SpannerException) ee.getCause();
        if (e.getErrorCode() != ErrorCode.ALREADY_EXISTS) {
          throw e;
        }
      }
      try {
        spanner.getDatabaseAdminClient().createDatabase("test-instance", "test-db", Arrays.asList("CREATE TABLE Singers (\n"
            + "  SingerId  INT64 NOT NULL,\n"
            + "  FirstName STRING(200),\n"
            + "  LastName  STRING(200) NOT NULL,\n"
            + "  FullName  STRING(400) NOT NULL AS (COALESCE(FirstName || ' ', '') || LastName) STORED,\n"
            + "  BirthDate DATE,\n"
            + "  Picture   BYTES(MAX),\n"
            + ") PRIMARY KEY (SingerId)")).get();
      } catch (ExecutionException ee) {
        SpannerException e = (SpannerException) ee.getCause();
        if (e.getErrorCode() != ErrorCode.ALREADY_EXISTS) {
          throw e;
        }
      }
      
      DatabaseClient client = spanner.getDatabaseClient(
          DatabaseId.of("test-project", "test-instance", "test-db"));
      client.readWriteTransaction().run(new TransactionCallable<Long>() {
        @Override
        public Long run(TransactionContext transaction) throws Exception {
          return transaction.executeUpdate(Statement
              .newBuilder("INSERT INTO Singers (SingerId, FirstName, LastName) VALUES (@id, @first, @last)")
              .bind("id").to(1L)
              .bind("first").to("Pete")
              .bind("last").to("Harrisson")
              .build()
              );
        }
      });
    }
  }
```

Error:
```
Exception in thread "main" com.google.cloud.spanner.SpannerException: UNKNOWN: com.google.api.gax.rpc.UnknownException: io.grpc.StatusRuntimeException: UNKNOWN: Unexpected error in RPC handling
	at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerExceptionPreformatted(SpannerExceptionFactory.java:259)
	at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerException(SpannerExceptionFactory.java:59)
	at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerException(SpannerExceptionFactory.java:179)
	at com.google.cloud.spanner.spi.v1.GapicSpannerRpc.get(GapicSpannerRpc.java:1665)
	at com.google.cloud.spanner.spi.v1.GapicSpannerRpc.executeQuery(GapicSpannerRpc.java:1417)
	at com.google.cloud.spanner.TransactionRunnerImpl$TransactionContextImpl.executeUpdate(TransactionRunnerImpl.java:595)
	at com.google.cloud.spanner.PdmlRunner$1.run(PdmlRunner.java:58)
	at com.google.cloud.spanner.PdmlRunner$1.run(PdmlRunner.java:1)
	at com.google.cloud.spanner.TransactionRunnerImpl$1.call(TransactionRunnerImpl.java:888)
	at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:105)
	at com.google.cloud.RetryHelper.run(RetryHelper.java:76)
	at com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:50)
	at com.google.cloud.spanner.SpannerRetryHelper.runTxWithRetriesOnAborted(SpannerRetryHelper.java:65)
	at com.google.cloud.spanner.SpannerRetryHelper.runTxWithRetriesOnAborted(SpannerRetryHelper.java:56)
	at com.google.cloud.spanner.TransactionRunnerImpl.runInternal(TransactionRunnerImpl.java:948)
	at com.google.cloud.spanner.TransactionRunnerImpl.run(TransactionRunnerImpl.java:847)
	at com.google.cloud.spanner.SessionPool$SessionPoolTransactionRunner.run(SessionPool.java:936)
	at com.google.cloud.spanner.PdmlRunner.main(PdmlRunner.java:55)
Caused by: java.util.concurrent.ExecutionException: com.google.api.gax.rpc.UnknownException: io.grpc.StatusRuntimeException: UNKNOWN: Unexpected error in RPC handling
	at com.google.common.util.concurrent.AbstractFuture.getDoneValue(AbstractFuture.java:566)
	at com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:547)
	at com.google.cloud.spanner.spi.v1.GapicSpannerRpc.get(GapicSpannerRpc.java:1659)
	... 14 more
Caused by: com.google.api.gax.rpc.UnknownException: io.grpc.StatusRuntimeException: UNKNOWN: Unexpected error in RPC handling
	at com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:47)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:72)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:60)
	at com.google.api.gax.grpc.GrpcExceptionCallable$ExceptionTransformingFuture.onFailure(GrpcExceptionCallable.java:97)
	at com.google.api.core.ApiFutures$1.onFailure(ApiFutures.java:68)
	at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1041)
	at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:30)
	at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1215)
	at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:983)
	at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:771)
	at io.grpc.stub.ClientCalls$GrpcFuture.setException(ClientCalls.java:563)
	at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:533)
	at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
	at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
	at com.google.cloud.spanner.spi.v1.SpannerErrorInterceptor$1$1.onClose(SpannerErrorInterceptor.java:100)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:617)
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:70)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:803)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:782)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: io.grpc.StatusRuntimeException: UNKNOWN: Unexpected error in RPC handling
	at io.grpc.Status.asRuntimeException(Status.java:533)
	... 18 more
```

I initially ran into this issue while working on the C# client. That client uses `ExecuteStreamingSql` to execute DML. I then tried it with the Java client which uses `ExecuteSql` to execute DML, so it does not seem to be specific for one of those methods, and also not to one specific client language.

**Comment by chao-tian** [comment]
Thanks for the additional info. I copied your java code and ran it locally. Unfortunately, I was unable to reproduce your issue. Could you check the following?
- Can you reproduce the error without any generated columns?
- Is there any emulator process from previous runs that are still running in the background. If you're on Linux, run "ps aux | grep emulator".

**Comment by olavloite** [comment]
This morning I'm working on the C# client, so I'm currently on Windows:
* No, so far I've only been able to reproduce the error using a generated column. It is however somewhat random. This morning the problem did not occur before I had executed a couple of tests, and it is currently occurring in about 30-40% of the insert statements on a table with a computed column.
* No, there are no emulator processes lingering from previous runs that I can find.

I just executed a `docker image prune -a` and downloaded the latest emulator again. That seems to have reduced the probability of the error happening somewhat, but I still get the (an) error in something like 10% of the cases. Sometimes the aforementioned error, and sometimes it crashes the emulator.

The last time the emulator crashed, it returned the error

```---- Grpc.Core.RpcException : Status(StatusCode="Unknown", Detail="Stream removed", DebugException="Grpc.Core.Internal.CoreErrorDetailException: {"created":"@1613643380.128000000","description":"Error received from peer ipv6:[::1]:9010","file":"T:\src\github\grpc\workspace_csharp_ext_windows_x64\src\core\lib\surface\call.cc","file_line":1062,"grpc_message":"Stream removed","grpc_status":2}")```

and then terminates.

**Comment by chao-tian** [comment]
Thanks! I am able to reproduce the bug now. It happens when there are regular columns defined after the generated columns in the same table. If you move the FullName column to the end of the table, the emulator should work fine.

We will fix the bug ASAP. And thanks again for reporting this bug.

**Comment by olavloite** [comment]
> Thanks! I am able to reproduce the bug now. It happens when there are regular columns defined after the generated columns in the same table. If you move the FullName column to the end of the table, the emulator should work fine.
> 
> We will fix the bug ASAP. And thanks again for reporting this bug.

Ah, great, thanks looking into this so quickly! I'll apply the workaround for now.

**Comment by jacobdufault-google** [comment]
The fix for this is included in the 1.3 release (738d9cc62038a84981f749acf9b43d86c72fbf12).
